src/doc/algorithm/shift_image.cpp

```diff
@@ -20,31 +23,111 @@
+  int maxY = MAX(cel->bounds().y+cel->bounds().h, mask->bounds().y+mask->bounds().h);
+  gfx::Rect compCelBounds = gfx::Rect(minX, minY, maxX - minX, maxY - minY);
+  Image* compImage = Image::create(cel->image()->pixelFormat(), compCelBounds.w, compCelBounds.h);
+
+  // Making a Rect which represents the mask bounds of the original MASK relative to
+  // the new COMPOUND IMAGE (compImage)
+  gfx::Point maskCelGap(0, 0);
+  if (cel->bounds().x < mask->bounds().x)
+    maskCelGap.x = mask->bounds().x - cel->bounds().x;
+  else
+    maskCelGap.x = 0;
+  if (cel->bounds().y < mask->bounds().y)
+    maskCelGap.y = mask->bounds().y - cel->bounds().y;
+  else
+    maskCelGap.y = 0;
+
+  gfx::Rect maskedBounds(maskCelGap.x,
+                         maskCelGap.y,
+                         mask->bounds().w,
+                         mask->bounds().h);
+
+  // Making one combined image: Image with the Mask Bounds (unfilled spaces were filled with mask color)
+  compImage->copy(cel->image(), gfx::Clip(cel->position().x-compCelBounds.x,
+                                          cel->position().y-compCelBounds.y,
+                                          0, 0,
+                                          cel->bounds().w, cel->bounds().h));
+
+  // Making a copy of only the image which will be shiftted
+  Image* imageToShift = Image::create(compImage->pixelFormat(), maskedBounds.w, maskedBounds.h);
+  imageToShift->copy(compImage, gfx::Clip(0, 0,
+                                          maskedBounds.x,
+                                          maskedBounds.y,
+                                          imageToShift->bounds().w, imageToShift->bounds().h));
+
+  // Shiftting the masked area of the COMPOUND IMAGE (compImage).
+  int initialX = maskedBounds.x;
+  int initialY = maskedBounds.y;
+  int finalX = maskedBounds.x2();
+  int finalY = maskedBounds.y2();
+  for (int y=initialY; y<finalY; ++y) {
+    for (int x=initialX; x<finalX; ++x) {
+        put_pixel(compImage,
+                  initialX + (maskedBounds.w + dx + x-initialX) % maskedBounds.w,
+                  initialY + (maskedBounds.h + dy + y-initialY) % maskedBounds.h,
+                  get_pixel(imageToShift, x - initialX, y - initialY));
+    }
+  }
+
+  // Bounds and Image shrinking (we have to fit compound image (compImage) and bounds (compCelBounds))
+  gfx::Rect newBounds(0, 0, compImage->bounds().w, compImage->bounds().h);
+  if (algorithm::shrink_bounds(compImage, newBounds, compImage->maskColor())) {
+    compCelBounds.offset(newBounds.x, newBounds.y);
+    compCelBounds.setSize(newBounds.size());
+  }
+  Image* finalImage = Image::create(compImage->pixelFormat(), compCelBounds.w, compCelBounds.h);
+  finalImage->copy(compImage, gfx::Clip(0, 0, newBounds.x, newBounds.y,
+                                        compCelBounds.w, compCelBounds.h));
+  
+  // Final cel content assign
+  finalImage->incrementVersion();
```

Comment:
Generally we would prefer all `incrementVersion()` calls to be inside `app::cmd` classes, but there are some exceptions. Just add a comment:
```suggestion
  finalImage->incrementVersion(); // TODO this should be in app::cmd module
```